TRPOD 312 - TPROD 358 - CoreBundle Theme: migrate views from php engine to twig

LordRembo opened the pull request:
<!-- ## Which branch should I use for my PR?

Assuming that:

a = current major release
b = current minor release
c = future major release

* a.x for any features and enhancements (e.g. 5.x)
* a.b for any bug fixes (e.g. 4.4, 5.1)
* c.x for any features, enhancements or bug fixes with backward compatibility breaking changes (e.g. 5.x) -->

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | [ ]
| Deprecations?                          | [ ]
| BC breaks? (use the c.x branch)        | [ ]
| Automated tests included?              | [ ] <!-- All PRs must maintain or improve code coverage -->
| Related user documentation PR URL      | mautic/mautic-documentation#... <!-- required for new features -->
| Related developer documentation PR URL | mautic/developer-documentation#... <!-- required for developer-facing changes -->
| Issue(s) addressed                     | Fixes #... <!-- prefix each issue number with "Fixes #", no need to create an issue if none exists, explain below instead -->

<!--
Additionally (see https://contribute.mautic.org/contributing-to-mautic/developer/code/pull-requests#work-on-your-pull-request):
 - Always add tests and ensure they pass.
 - Bug fixes must be submitted against the lowest maintained branch where they apply
   (lowest branches are regularly merged to upper ones so they get the fixes too.)
 - Features and deprecations must be submitted against the "4.x" branch.
-->

#### Description:
- migrated template(s) responsible for the Themes overview (/s/themes)
- in that process, I also had to tweak some related templates from Default (eg. content.html.twig, base.html.twig & slim.html.twig, …).
- Making these changes, meant switching to the twig templates instead of php templates for Left panel, Right panel, global search and notifications panel on some pages that did not use them yet.
- The impact is on pages that use the same 'list' component as the Themes overview (eg. Segments overview)
- Also impacts their detail pages (the View & Edit), as these inherit the same base templates to load the global search and the actions (buttons on top right).

<!--
Please write a short README for your feature/bugfix. This will help people understand your PR and what it aims to do. If you are fixing a bug and if there is no linked issue already, please provide steps to reproduce the issue here.
-->

#### Steps to test this PR:
- go to /s/themes:
  - you should see the list of themes, with title, actions, … same as before
  - you can still use the actions to download or preview
  - left & right sidebar should still show menu items and work when clicked
  -  clicking search icon in header should still open up search and give you the expected results
  - notifications icon should still open up notifications pane
- other first-level navigation items that contain lists, should also still show their list and you should be able to click through to detail. These should also have working sidebars, search and notifications pane.
For example:
  - /s/contacts
  - /s/companies
  - /s/segments
  - s/forms
  - s/pages
- for good measure, you should also check the detail and Edit view of one of the contacts, segments, companies. This to make sure the buttons display and work, search works, sidebars work

<!--
This part is really important. If you want your PR to be merged, take the time to write very clear, annotated and step by step test instructions. Do not assume any previous knowledge - testers may not be developers.
-->
1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. 

<!--
If you have any deprecations, list them here along with the new alternative.
If you have any backwards compatibility breaks, list them here.
-->

LordRembo committed:
Make Theme page use the twig files, rework part of the inheritance of twig theme files

LordRembo committed:
get rid of chain of includes in the original twig and do extends instead, as we want to follow the php logic instead of refactoring & adding complexity
also converted some blocks into variables for same reason

LordRembo committed:
expanded the array check a bit for clarity

LordRembo committed:
fix issue with delete button and filter the correct themes from the list

LordRembo committed:
added a twig template for form (used in the header for the upload field part)

LordRembo committed:
bit of cleanup

LordRembo committed:
make pageHeader and head component load the needed variables by passing them on earlier+ clean up the vars a bit and get rid of last instance of headerTitle block (is now simply a var)

LordRembo committed:
rename 'panel' block to 'content' block, as that is the standard name used elsewhere in templates

LordRembo committed:
set default content for the '_content' block (sidenote: best rename this block later)

LordRembo committed:
cleanup: remove a redundant space

LordRembo commented:
Got a problem with another test from Leadbundle. Will take look and see if it's again to do with html encoding

RCheesley reviewed (changes requested):
Things I noticed:

- In the list view of themes the URLs are not seemingly being correctly rendered into links:

`https://8080-lordrembo-mautic-ggws5ib3tcx.ws-eu86.gitpod.io/Mautic/CoreBundle/Helper/InputHelper::url(https://mautic.org);%20?%3E`

- It seems to take a _very_ long time to open the menu items and when opening there is some code that flashed at the top of the screen: https://watch.screencastify.com/v/rvX9WZ3JegELGSZTYLb8 - this is much more noticeable on contact profile pages.
- Global search works but errors are being shown: https://watch.screencastify.com/v/0Q4Qc18Czd4jZ6PJyU9b but nothing is reported in the log files
- We seem to be missing some page titles, I noticed:

The contact name is missing from here:

![screenshot-8080-lordrembo-mautic-ggws5ib3tcx ws-eu86 gitpod io-2023 02 14-20_21_43](https://user-images.githubusercontent.com/2930593/218853217-2ecc0f84-adef-4109-8444-4133e0fcde4b.png)

Also the company name:

![screenshot-8080-lordrembo-mautic-ggws5ib3tcx ws-eu86 gitpod io-2023 02 14-20_22_55](https://user-images.githubusercontent.com/2930593/218853649-2caf89b8-afd4-42b4-91ce-a02af3569b84.png)

Also missing Segments title here:

![screenshot-8080-lordrembo-mautic-ggws5ib3tcx ws-eu86 gitpod io-2023 02 14-20_24_14 (1)](https://user-images.githubusercontent.com/2930593/218853913-83930941-e01a-4348-8848-97854d1b6bf5.png)

LordRembo committed:
cleanup Themes form template

LordRembo commented:
Dangit, I was afraid of things like this. I've got a pretty good idea where/how those titles are supposed to load so that should be relatively easy to fix. The failed unit tests are probably related to the same component(s) used for the missing titles.
The code flashing and the errors, on the other hand, are more annoying to trace. Guess I know what to do on Thursday :D

LordRembo committed:
Revert pageTitle logic to using block instead of just variable (since this system is used in other bundles as well)

LordRembo committed:
Fix url output in Themes list items

LordRembo committed:
revert Default content html twig template to previous state, as ajax calls are breaking across other bundles (refactor this later)

LordRembo committed:
set some defaults across various templates to avoid untracable 500 errors

LordRembo committed:
fix typo in a variable that is printed in Standard list

LordRembo committed:
set default pageTitle and headerTitle in Error and Exception templates, so Ajax 404 errors don't fail and fall back to a 500 error

LordRembo commented:
- Fixed urls in Theme listing
- Fixed missing page titles
- fixed glitch where on some pages raw data was printed when going to another page
- ~Global search should be fixed as well, as this was related to same issue: wrong template was used when fetching content via Ajax~ The search error is apparently [an issue unrelated to this PR](https://github.com/mautic/mautic/issues/11977)

LordRembo commented:
~Looking into the failed test and some weirdness with notifications.~  all fixed now

LordRembo committed:
Re-run tests

LordRembo committed:
don't pass the variables to the default output include in base tempates (no need + breaks inheritance)

LordRembo committed:
don't pass the variables to the pageheader in default output since they are passed along anyway (and we removed this in extended templates as well)

LordRembo committed:
don't output raw actions when it's not html

LordRembo committed:
Merge branch '5.x' into TPROD-312/TPROD-358-corebundle-theme-migrate-views-from-php-engine-to-twig

sdoering commented:
Hi, 

just tested it. 
- [x] Downloading a theme works
- [x] Uploading a theme works

Not sure if the search is an issue, as it does not list the themes in its results (basically it doesn't find them): 
![image](https://user-images.githubusercontent.com/765204/219851238-70a3e796-e59d-4a26-9695-996962a3cca1.png)

- [x] left & right sidebar should still show menu items and work when clicked
- [x] notifications icon should still open up notifications pane

Everything else I tested is still working as expected.

LordRembo committed:
Merge branch '5.x' into TPROD-312/TPROD-358-corebundle-theme-migrate-views-from-php-engine-to-twig

LordRembo commented:
Oh, that is strange indeed. I don't think the changes in this PR should influence the results of a search, as those templates and that logic are separate. I can double-check on Monday and compare with the 5.x branch if it happens there as well.
Would be helpful if anyone else can check if they see that issue here (or on 5.x)

mabumusa1 commented:
> Oh, that is strange indeed. I don't think the changes in this PR should influence the results of a search, as those templates and that logic are separate. I can double-check on Monday and compare with the 5.x branch if it happens there as well. Would be helpful if anyone else can check if they see that issue here (or on 5.x)

I do not think this is related to this PR, I searched for a theme, and did not show on branch 5.x, so I think it is not related to this PR

LordRembo committed:
Merge branch '5.x' into TPROD-312/TPROD-358-corebundle-theme-migrate-views-from-php-engine-to-twig

mabumusa1 reviewed (approved):
I tested this PR, works as expected

RCheesley reviewed (approved):
Great job @LordRembo - all issues I had found are fixed in all respects, able to upload and work with themes. 

Thanks for your work on this PR, LGTM! 🚀

escopecz commented on app/bundles/CoreBundle/Views/Helper/list_actions.html.twig:
```diff
@@ -8,7 +8,7 @@
 {{ restore('action_button_helper') }}
 
-{% if item is iterable %}
+{% if item is iterable and item[0] is not defined %}
```
I'm not a fan of this. The item array does not have to start with 0. Sometimes the IDs from the database are used as array keys. Can this check for the actual property it needs?
```twig
{% if item is iterable and item.id is defined and item.name is defined %}
```

LordRembo replied:
I want to check if `item` is an object and not an array. This particular line mimics what PHP does, only Twig doesn't have a complete check like php does. I think this is much cleaner than checking for particular keys.

LordRembo replied:
There's no reason I can see that the item would contain array keys. Can you give me an example?

LordRembo replied:
Ah, hold on, you're right. I was confusing this check with one I got rid of in another template. My bad, sorry!
Wil make changes and set up a new PR. It will have to wait until Thursday though, as I won't be working tomorrow (Wednesday)
Thanks for catching this!

LordRembo replied:
Made a new PR for the changes in this review

LordRembo replied:
[Made a new PR for the changes in this review](https://github.com/mautic/mautic/pull/12017)

escopecz commented on app/bundles/CoreBundle/Views/Theme/form.html.twig:
```diff
@@ -0,0 +1,22 @@
+{#
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+#}
```
Remove, please

LordRembo replied:
Any particular reason this needs to be removed? The PHP templates also carry meta-information like this.

escopecz commented on app/bundles/CoreBundle/Views/Theme/index.html.twig:
```diff
@@ -0,0 +1,35 @@
+{#
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+#}
+
```
Remove, please

LordRembo replied:
Any particular reason this needs to be removed? The PHP templates also carry meta-information like this.

escopecz commented on app/bundles/CoreBundle/Views/Theme/list.html.twig:
```diff
@@ -0,0 +1,153 @@
+{#
+/*
+ * @copyright   2014 Mautic Contributors. All rights reserved
+ * @author      Mautic
+ *
+ * @link        http://mautic.org
+ *
+ * @license     GNU/GPLv3 http://www.gnu.org/licenses/gpl-3.0.html
+ */
+ #}
```
Remove, please

LordRembo replied:
Any particular reason this needs to be removed? The PHP templates also carry meta-information like this.

escopecz replied:
Please see https://github.com/mautic/mautic/pull/10970 for the reasons

escopecz reviewed (commented):
Aha, that's why there are some features of the review UI missing. It's merged already.

mabumusa1 commented:
> Aha, that's why there are some features of the review UI missing. It's merged already.

It was my mistake, I merged it before you did your review. 

@LordRembo may you please start a new PR with @escopecz feedback, I merged this PR before resolving his comments.

Sorry for the inconvenience

LordRembo commented:
No problem. Will follow up on these comments.

LordRembo commented:
Follow-up in this PR: https://github.com/mautic/mautic/pull/12017
